Loki: Use label/<name>/values API instead of series API for label values discovery

yuri-rs opened the pull request:
**What is this feature?**

Loki: Add label values API selector to setting and use it when selected

**Why do we need this feature?**

Make loki label_values more efficient

**Who is this feature for?**

Loki users

**Which issue(s) does this PR fix?**:

Fixes #74432 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

yuri-rs committed:
Loki: Add label values API selector to setting and use label_values API when selected

ifrost commented on CHANGELOG.md:
```diff
@@ -23,6 +23,7 @@ To resolve a technical issue within the Grafana release package management proce
 
 ### Features and enhancements
 
+- **Loki:** Add label values API selector to setting and use label_values API when selected. [TBD], [@yuri-rs](https://github.com/yuri-rs)
```
How did it get here? 🤔  :D

yuri-rs replied:
removed

yuri-rs committed:
remove CHANGELOG change

yuri-rs committed:
add docs

ivanahuckova reviewed (commented):
Hello @yuri-rs, thank you for your contribution. However, I don't believe adding this extra setting to the data source configuration page is necessary. Instead, we could determine the Loki HTTP API endpoint to use based on the Loki version.

Currently, we have an open issue to figure out how to handle new features and versions in Loki, but it hasn't been prioritized yet. You can find more details about it here: https://github.com/grafana/grafana/issues/67626. However, until we resolve that, I am not sure we can move forward with this.

@grafana/observability-logs feel free to chime in if you have any different ideas.

yuri-rs commented:
> Hello @yuri-rs, thank you for your contribution. However, I don't believe adding this extra setting to the data source configuration page is necessary. Instead, we could determine the Loki HTTP API endpoint to use based on the Loki version.
> 
> Currently, we have an open issue to figure out how to handle new features and versions in Loki, but it hasn't been prioritized yet. You can find more details about it here: #67626. However, until we resolve that, I am not sure we can move forward with this.
> 
> @grafana/observability-logs feel free to chime in if you have any different ideas.

Hello @ivanahuckova,

Thank you for the feedback and for linking to the discussion about handling new features and versions in Loki.
I have experience with the Prometheus plugin and the method used for selecting the Prometheus type (Prometheus|Mimir|Thanos) version, and it seems much less intuitive. I recommend not adopting this approach for the Loki plugin.
In my opinion, the optimal solution would be to introduce an official logql endpoint that provides a list of features supported by the datastore.
This approach would enable future logql-compatible datastores (like Mimir for Prometheus) to utilize the Loki plugin without plugin changes.

yuri-rs commented:
This PR is still valid. Please don't close it.

@ivanahuckova As I see it, automatic feature detection in the Loki plugin could be a long-drawn process, but the changes I propose would benefit many users.
Could we proceed with manual API selection in the data source right now and transition to automatic detection in the future?
I'm sure users would be happy to have a way to use the labels API, even if it requires manual selection in the data source.

ivanahuckova commented:
Hi @yuri-rs, we've had a discussion with the team, and considering that this feature was added in Loki v2.7, and in the next Grafana release we will be [supporting Loki v2.8+](https://grafana.com/docs/grafana/next/datasources/loki/#supported-loki-versions), we can implement this improvement. Moreover, we don't need to add a `Label match API` config to the Loki data source or access the `/labels` API.

Are you okay with removing config, or do you want me to implement the changes?

yuri-rs committed:
Merge branch 'main' into loki/labels-match-api-support

yuri-rs commented:
Hi @ivanahuckova,
I will update the PR to unconditionally use new /labels API.

yuri-rs committed:
Support Loki 2.7+ only, so unconditionally use /labels API

yuri-rs committed:
Correct doc for fetchLabels and fetchLabelValues functions

yuri-rs commented:
Hi @ivanahuckova,
All done, please take a look. It works fine locally for me.

matyax commented:
Sure!

matyax reviewed (dismissed):
The changes look good, but there seems to be an issue while autocompleting that results in a visible error. Maybe there's a missing error handling somewhere. Unfortunately I don't have capacity to dig deeper or fix.


https://github.com/grafana/grafana/assets/1069378/75a4bd7e-1e43-41da-a441-93ffb0fd923b

yuri-rs committed:
Merge branch 'grafana:main' into loki/labels-match-api-support

yuri-rs committed:
Fixes after merge

yuri-rs committed:
Do not encode query parameter twice

yuri-rs commented:
@matyax Thanks for the review. I have fixed the problem.
The issue was in request "query" parameter encoding, it was encoded twice, so `{}` became `%7B%7D` and then encoded further to `%257B%257D`.
It works that way even before my PR, but was hidden since label API was never used with stream selector. So, I removed an extra params encoding. 
I have tested the new version of PR within Grafana cloud Loki and it works for me now.
Could you please have another look at the PR?

matyax committed:
Merge branch 'main' into loki/labels-match-api-support

matyax commented:
I should have resolved the conflicts. If not, please fix the CI checks and I'll review next week. Thanks!

ivanahuckova commented on public/app/plugins/datasource/loki/LanguageProvider.ts:
```diff
@@ -229,9 +251,11 @@ export default class LokiLanguageProvider extends LanguageProvider {
-      : undefined;
+    // Loki don't allow empty streamSelector {}, so we should not send it.
+    const streamParam =
+      options?.streamSelector && options.streamSelector !== '{}'
```
nit pick, I would extract `'{}'` to a constant (e.g. `const EMPTY_SELECTOR = '{}'`) and used that across codebase.

yuri-rs replied:
fixed

ivanahuckova commented on public/app/plugins/datasource/loki/LanguageProvider.ts:
```diff
@@ -229,9 +251,11 @@ export default class LokiLanguageProvider extends LanguageProvider {
-    const streamParam = options?.streamSelector
-      ? encodeURIComponent(this.datasource.interpolateString(options.streamSelector))
-      : undefined;
+    // Loki don't allow empty streamSelector {}, so we should not send it.
```
```suggestion
    // Loki doesn't allow empty streamSelector {}, so we should not send it.
```

yuri-rs replied:
fixed, thanks

ivanahuckova commented on public/app/plugins/datasource/loki/LanguageProvider.ts:
```diff
@@ -118,6 +118,28 @@ export default class LokiLanguageProvider extends LanguageProvider {
+   */
+  async fetchLabels(options?: { streamSelector?: string; timeRange?: TimeRange }): Promise<string[]> {
+    // If there is no stream selector - use /labels endpoint (https://github.com/grafana/loki/pull/11982)
+    if (!options || !options.streamSelector) {
```
Here we have the same issue with empty stream selector, so we need to add a check - something like: 
```suggestion
    if (!options?.streamSelector || options.streamSelector === EMPTY_SELECTOR) {
```

yuri-rs replied:
Since Loki series endpoint supports empty match[] it is fine to use series endpoint for fetchLabels calls with defined, but empty seriesSelector.
What you propose is just a micro-optimization.
Do you want me to change it anyway?

ivanahuckova replied:
👍 thanks for confirming that series endpoint supports `{}`. It is fine then, no need to change.

ivanahuckova commented on public/app/plugins/datasource/loki/LanguageProvider.ts:
```diff
@@ -118,6 +118,28 @@ export default class LokiLanguageProvider extends LanguageProvider {
+   * @returns A promise containing an array of label values.
+   * @throws An error if the fetch operation fails.
+   */
+  async fetchLabels(options?: { streamSelector?: string; timeRange?: TimeRange }): Promise<string[]> {
```
When we are adding new data source APIs, we also document them in `docs/app_plugin_developer_documentation.md`

yuri-rs replied:
Thanks, missed that. fixed

ivanahuckova commented on public/app/plugins/datasource/loki/LanguageProvider.test.ts:
```diff
@@ -171,7 +171,7 @@ describe('Language completion provider', () => {
       expect(requestSpy).toHaveBeenCalledWith('label/testkey/values', {
         end: 1560163909000,
-        query: '%7Bfoo%3D%22bar%22%7D',
+        query: '{foo="bar"}',
```
I would prefer if we didn't change the encoding in this PR as it is not related to the changes, and if you'd like to do it, then change it in separate PR.

ivanahuckova commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:
```diff
@@ -39,24 +39,19 @@ export class CompletionDataProvider {
   }
 
   async getLabelNames(otherLabels: Label[] = []) {
-    if (otherLabels.length === 0) {
-      // if there is no filtering, we have to use a special endpoint
-      return this.languageProvider.getLabelKeys();
-    }
```
Here we want to keep `this.languageProvider.getLabelKeys()`; as it provides cached loaded/cached labels, while your implementation re-requests these labels.

yuri-rs replied:
fixed, thanks

ivanahuckova commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:
```diff
@@ -39,24 +39,19 @@ export class CompletionDataProvider {
-
-    const data = await this.getSeriesLabels(otherLabels);
-    return data[labelName] ?? [];
+    return await this.languageProvider.fetchLabelValues(labelName, {
```
🤩

ivanahuckova commented on public/app/plugins/datasource/loki/datasource.ts:
```diff
@@ -680,16 +680,10 @@ export class LokiDatasource
-    }
-
-    const result = await this.languageProvider.fetchLabelValues(query.label, { timeRange });
+    const result = await this.languageProvider.fetchLabelValues(query.label, {
```
🤩

ivanahuckova reviewed (changes requested):
Hi @yuri-rs! First of all, I apologize for the delay in our review; we were busy with G11 and Explore Logs app. From now on, reviews should proceed much faster. The PR looks great. Thank you so much for making the effort to update and refactor all the places where the code could be simplified. I have left some feedback and requested a few changes.

yuri-rs committed:
return getLabelKeys in Completion Provider

yuri-rs committed:
docs

yuri-rs committed:
Add test for LabelParamEditor

yuri-rs commented:
> Hi @yuri-rs! First of all, I apologize for the delay in our review; we were busy with G11 and Explore Logs app. From now on, reviews should proceed much faster. The PR looks great. Thank you so much for making the effort to update and refactor all the places where the code could be simplified. I have left some feedback and requested a few changes.

Hi @ivanahuckova,
Thanks for the review. I have fixed review comments and added missed test.
Please take a look again

ivanahuckova commented on public/app/plugins/datasource/loki/LogContextProvider.test.ts:
```diff
@@ -425,14 +425,14 @@ describe('LogContextProvider', () => {
       });
 
-      it('should call fetchSeriesLabels if parser', async () => {
+      it('sshould call fetchLabels with stream selector if parser', async () => {
```
```suggestion
      it('should call fetchLabels with stream selector if parser', async () => {
```

ivanahuckova reviewed (commented):
LGTM! But you still need to resolve the merge conflicts. After they are correctly resolved I can approve and merge. 🙂

yuri-rs committed:
Update public/app/plugins/datasource/loki/LogContextProvider.test.ts

Co-authored-by: Ivana Huckova <30407135+ivanahuckova@users.noreply.github.com>

yuri-rs committed:
Merge branch 'main' into loki/labels-match-api-support

yuri-rs committed:
toHaveBeenCalledWith

ivanahuckova commented:
Merged @yuri-rs. Thanks again SO MUCH for this contribution! 🫶

svennergr commented:
Thank you @yuri-rs!
